### PR TITLE
Addressing some TODOs in non-clifford functionality along with resolving #309

### LINKED ISSUE #309: Completing the non-Clifford capabilities [$800]
<details>
<summary><strong>Bug bounty logistic details</strong> (click to expand)</summary>

To claim exclusive time to work on this bounty either post a comment here or message [skrastanov@umass.edu](mailto:skrastanov@umass.edu) with:

- your name
- github username
- **(optional)** a brief list of previous pertinent projects you have engaged in

If you want to, you can work on this project without making a claim, however claims are encouraged to give you and other contributors peace of mind. Whoever has made a claim takes precedence when solutions are considered.

You can always propose your own funded project, if you would like to contribute something of value that is not yet covered by an official bounty.
</details>

The project is claimed by @Fe-r-oz  until Sep 10th 2024.

# Project: Completing the non-Clifford capabilities [$800]

This library already has some nascent capabilities to represent non-Clifford states as a weighted sum of tableaux states. Such a representation is useful in settings where there is a small number of non-Clifford gates in a mostly Clifford circuit. The cost is still exponential in the number of non-Clifford gates, but manageable. We have simple state representation and gate application already done, `expect` is [nearly implemented in this PR](https://github.com/QuantumSavory/QuantumClifford.jl/pull/259), and `project!` still needs to be implemented. To obtain this bounty, the aforementioned features need to be completed, well tested, and well documented.

**Required skills**: Understanding of the Stabilizer formalism and decompositions of arbitrary states in terms of Pauli channels

**Reviewer**: Stefan Krastanov

**Duration**: 2 months

**Publication**: In the next 2 years we plan to release a paper in a selective journal about this software. Contributing to this issue would deserve a co-authorship status on such a paper (if the contributor so desires)

#### Payout procedure:

The Funding for these bounties comes from the National Science Foundation and from the NSF Center for Quantum Networks. The payouts are managed by the NumFOCUS foundation and processed in bulk once every two months. If you live in a country in which NumFOCUS can make payments, you can participate in this bounty program.

[Click here for more details about the bug bounty program.](https://github.com/QuantumSavory/.github/blob/main/BUG_BOUNTIES.md)

### PR DESCRIPTION
**Pull request by Fe-r-oz**
This PR aims to address the TODOs that are provided in the non-clifford functionality and further resolve #309.

Sorry for inconvenience, @Krastanov, I will have to learn about interactive rebase stuff. I wanted to have a chat that I address some of the comments that you had in the form of TODOs because I was checking out the #259 and becoming aware of the cool work on the non-clifford. But I was not sure what to do, either ask you as a question, or comments it in the PR, so I decided to have the improvements as a comment. But, as you suggested, better to have PR.

I am going through your implementation and saw some TODOs, so I decided that implementing them will make me more familiar with the functionality to understand the changes that needs to be introduced to resolve Issue 309 

Before appling improvement to TODOs:
```
"""Same as `all(==(0), (a.+b.+c) .% 2)`"""
function _allthreesumtozero(a,b,c) # TODO consider using bitpacking and SIMD xor with less eager shortcircuiting -- probably would be much faster
    @inbounds @simd for i in 1:length(a)
        iseven(a[i]+b[i]+c[i]) || return false
    end
    true
end

function _dictvaltype(dict)
    return eltype(dict).parameters[2] # TODO there must be a cleaner way to do this
end

function apply!(state::GeneralizedStabilizer, gate::PauliChannel)
    dict = state.destabweights
    stab = state.stab
    dtype = _dictvaltype(dict)
    tzero = zero(dtype)
    tone = one(dtype)
    newdict = typeof(dict)(tzero) # TODO jeez, this is ugly
    for ((dᵢ,dⱼ), χ) in dict # the state
        for ((Pₗ,Pᵣ), w) in zip(gate.paulis,gate.weights) # the channel
            phaseₗ, dₗ, dₗˢᵗᵃᵇ = rowdecompose(Pₗ,stab)
            phaseᵣ, dᵣ, dᵣˢᵗᵃᵇ = rowdecompose(Pᵣ,stab)
            c = (dot(dₗˢᵗᵃᵇ,dᵢ) + dot(dᵣˢᵗᵃᵇ,dⱼ))*2
            dᵢ′ = dₗ .⊻ dᵢ
            dⱼ′ = dᵣ .⊻ dⱼ
            χ′ = χ * w * (-tone)^c * (im)^(-phaseₗ+phaseᵣ+4)
            newdict[(dᵢ′,dⱼ′)] += χ′
        end
    end
    for (k,v) in newdict # TODO is it safe to modify a dict while iterating over it?
        if abs(v) < 1e-14 # TODO parameterize this pruning parameter
            delete!(newdict, k)
        end
    end
    state.destabweights = newdict
    state
end
```

After addressing the TODOs, please let me know are these improvements plausible.
```
"""Same as `all(==(0), (a.+b.+c) .% 2)`"""
function _allthreesumtozero(a, b, c)
    packed_sum = a .+ b .+ c
    all_even = all(@. iseven(packed_sum))
    return all_even
end

function _dictvaltype(dict)
    return Base.getindex(eltype(dict).parameters, 2)
end

function apply!(state::GeneralizedStabilizer, gate::PauliChannel)
    dict = state.destabweights
    stab = state.stab
    dtype = _dictvaltype(dict)
    tzero = zero(dtype)
    tone = one(dtype)
    newdict = DefaultDict{Tuple{BitVector, BitVector}, Complex{Float64}}(Complex{Float64}(0, 0))
    for ((dᵢ,dⱼ), χ) in dict
        for ((Pₗ,Pᵣ), w) in zip(gate.paulis, gate.weights)
            phaseₗ, dₗ, dₗˢᵗᵃᵇ = rowdecompose(Pₗ, stab)
            phaseᵣ, dᵣ, dᵣˢᵗᵃᵇ = rowdecompose(Pᵣ, stab)
            c = (dot(dₗˢᵗᵃᵇ, dᵢ) + dot(dᵣˢᵗᵃᵇ, dⱼ)) * 2
            dᵢ′ = dₗ .⊻ dᵢ
            dⱼ′ = dᵣ .⊻ dⱼ
            χ′ = χ * w * (-tone)^c * (im)^(-phaseₗ + phaseᵣ + 4)
            newdict[(dᵢ′, dⱼ′)] += χ′
        end
    end
    keys_to_delete = Tuple[]
    for (k, v) in newdict
        if abs(v) < 1e-14
            push!(keys_to_delete, k)
        end
    end
    for k in keys_to_delete
        delete!(newdict, k)
    end
    state.destabweights = newdict
    return state
end
```

I reran the this test to confirm that there was no error:
```

julia> for n in 1:5
           i = rand(1:n)
           stab = random_stabilizer(n)
           genstab = GeneralizedStabilizer(stab)
           ket = Ket(stab)
           @test dm(ket) ≈ Operator(stab)
           @test dm(ket) ≈ Operator(genstab)

           pauli = random_pauli(n; nophase=false, realphase=true)
           qo_pauli = Operator(pauli)

           qo_bigtgate = n==1 ? qo_tgate : embed(qo_basis^n, i, qo_tgate)
           bigtgate = embed(n,i, pcT)
           @test qo_bigtgate ≈ Operator(bigtgate)

           for step in 1:10
               # apply!(ket, qo_bigtgate) TODO implement this API
               ket = qo_bigtgate*ket
               apply!(genstab, bigtgate)
               @test dm(ket) ≈ Operator(genstab)
               @test isapprox(expect(qo_pauli, ket), expect(pauli, genstab); atol=1e-10)
           end
       end
```

### CODE REVIEW

**Comment by Krastanov** [comment]
Thanks! Here are a few more details on what I meant by "single commit = single semantic change" and how it relates to interactive rebase. Do not hesitate to ask if something is unclear! This workflow can be a bit annoying at first, but it really helps with organizing your work longterm.

I can look at the changes that you have made by clicking on your single new commit: https://github.com/QuantumSavory/QuantumClifford.jl/pull/313/commits/8a0f7781212aaa57cb938c6c6f02176e7c17abd9

I see three changes in it. They are small so it kinda makes sense for them to be in a single commit, but let's "role play" through a review:

### step 1
```
commits:
1. fixing three TODOs
```
I review and say "hey could you change something about the second TODO"

### step 2
you make that change and we have
```
commits:
1. fixing three TODOs
2. modifying the second TODO
```
Then we notice that there is something about the first todo and you fix that too and we end up with

### step 3
```
commits:
1. fixing three TODOs
2. modifying the second TODO
3. modifying the first todo
```
finally, there is something else that we notice about the second TODO and you fix that and we are left with

### step 4
```
commits:
1. fixing three TODOs
2. modifying the second TODO
3. modifying the first todo
4. modifying the second TODO again
```

So now if I want to review just the changes to the second TODO, I need to read through commits 1,2, and 4, which makes my work pretty inconvenient. Even worse, it is much more probable that I will miss something this way. I will not be able to think separately about each TODO, I will have to review them all together, risking missing something because I need to divine which change is related to which TODO.

Of course, this particular example is incredibly contrived, the changes are small, etc, so it is actually pretty easy for me to read all three TODOs together. You really do not need to change anything here. I am just using it as an example.

In this example, it would have been better if your commits looked like:
```
commits before review:
1. fixing the TODOs about yada yada (i.e. the first TODO)
2. fixing the TODOs about yada yada (i.e. the second TODO)
3. fixing the TODOs about yada yada (i.e. the third TODO)
commits after review:
5. modifying the second TODO
6. modifying the first todo
7. modifying the second TODO again
```

Again, this is just a contrived example, no need to actually do it here. Anyway, this would already be a (hypothetical) improvement. I can now select commits 2, 4, and 6 and look at the changes just to the second TODO. Importantly, at this point you should also run an interactive rebase `git rebase -i HEAD~n` (if you want to rebase only the last n commits) or `git rebase -i master` (if you want to rebase all commits in the pull request) and reorder the commits:

```
commits after rebase (just reordering)
1. fixing the TODOs about yada yada (i.e. the first TODO)
6. modifying the first todo
2. fixing the TODOs about yada yada (i.e. the second TODO)
5. modifying the second TODO
7. modifying the second TODO again
3. fixing the TODOs about yada yada (i.e. the third TODO)
```

And the commits should not only be reordered during the rebase, some of them should be squashed:

```
commits after rebase (including squashing)
1. fixing the TODOs about yada yada (i.e. the first TODO)
   squashing into 1: 6. modifying the first todo
2. fixing the TODOs about yada yada (i.e. the second TODO)
   squashed into 2: 5. modifying the second TODO
   squashed into 2: 7. modifying the second TODO again
3. fixing the TODOs about yada yada (i.e. the third TODO)
```

Now you would have only 3 neatly separated commits, each dedicated to a separate semantic change.

File: src/nonclifford.jl [hunk 1]
```diff
@@ -108,15 +108,14 @@ function expect(p::PauliOperator, s::GeneralizedStabilizer) # TODO optimize
 end
 
 """Same as `all(==(0), (a.+b.+c) .% 2)`"""
-function _allthreesumtozero(a,b,c) # TODO consider using bitpacking and SIMD xor with less eager shortcircuiting -- probably would be much faster
-    @inbounds @simd for i in 1:length(a)
-        iseven(a[i]+b[i]+c[i]) || return false
-    end
-    true
```
Comment (Krastanov) [inline src/nonclifford.jl]: Is this actually better? Superficially, I would expect this to be much slower and to allocate. Some benchmarks might be in order.

File: src/nonclifford.jl [hunk 2]
```diff
@@ -108,15 +108,14 @@ function expect(p::PauliOperator, s::GeneralizedStabilizer) # TODO optimize
...
+function _allthreesumtozero(a, b, c)
+    packed_sum = a .+ b .+ c
+    all_even = all(@. iseven(packed_sum))
+    return all_even
 end
 
 function _dictvaltype(dict)
-    return eltype(dict).parameters[2] # TODO there must be a cleaner way to do this
```
Comment (Krastanov) [inline src/nonclifford.jl]: This is going backwards. The `[]` is syntactic sugar for `getindex`. It does not make sense to go to the less readable `getindex`.

The original reason for this todo is because it is ugly and dangerous to rely on non-public APIs, specifically the `.parameters` access.

Looking around the Julia documentation, I see that there is `valtype` which is probably the correct way to fix this:

```
julia> valtype(Dict(:a=>1))
Int64
```

File: src/nonclifford.jl [hunk 3]
```diff
@@ -179,25 +178,29 @@ function apply!(state::GeneralizedStabilizer, gate::PauliChannel)
     dtype = _dictvaltype(dict)
     tzero = zero(dtype)
     tone = one(dtype)
-    newdict = typeof(dict)(tzero) # TODO jeez, this is ugly
-    for ((dᵢ,dⱼ), χ) in dict # the state
-        for ((Pₗ,Pᵣ), w) in zip(gate.paulis,gate.weights) # the channel
```
Comment (Krastanov) [inline src/nonclifford.jl:147]: removing these comments is not helpful -- these comments guide the reader what is being looped over

File: src/nonclifford.jl [hunk 4]
```diff
@@ -179,25 +178,29 @@ function apply!(state::GeneralizedStabilizer, gate::PauliChannel)
...
+    for (k, v) in newdict
+        if abs(v) < 1e-14
+            push!(keys_to_delete, k)
         end
     end
+    for k in keys_to_delete
+        delete!(newdict, k)
+    end
     state.destabweights = newdict
-    state
```
Comment (Krastanov) [inline src/nonclifford.jl:163]: Is this actually necessary? Is it actually dangerous to modify the dictionary keys while iterating over it?

File: src/nonclifford.jl [hunk 5]
```diff
@@ -179,25 +178,29 @@ function apply!(state::GeneralizedStabilizer, gate::PauliChannel)
...
             dᵢ′ = dₗ .⊻ dᵢ
             dⱼ′ = dᵣ .⊻ dⱼ
-            χ′ = χ * w * (-tone)^c * (im)^(-phaseₗ+phaseᵣ+4)
-            newdict[(dᵢ′,dⱼ′)] += χ′
+            χ′ = χ * w * (-tone)^c * (im)^(-phaseₗ + phaseᵣ + 4)
+            newdict[(dᵢ′, dⱼ′)] += χ′
         end
     end
-    for (k,v) in newdict # TODO is it safe to modify a dict while iterating over it?
-        if abs(v) < 1e-14 # TODO parameterize this pruning parameter
```
Comment (Krastanov) [inline src/nonclifford.jl:158]: removing the `parameterize this` TODO is unhelpful. At some point we might actually parameterize this -- we should keep the TODO to remind us of that unless it is fixed.

File: src/nonclifford.jl [hunk 6]
```diff
@@ -179,25 +178,29 @@ function apply!(state::GeneralizedStabilizer, gate::PauliChannel)
     dtype = _dictvaltype(dict)
     tzero = zero(dtype)
     tone = one(dtype)
-    newdict = typeof(dict)(tzero) # TODO jeez, this is ugly
```
Comment (Krastanov) [inline src/nonclifford.jl:145]: This is worse. Previously the type of the new dictionary was neatly dependent on the type of the input dictionary, ensuring generality. Now you have hardcoded the dictionary type (and you are not guaranteed that the rest of the code would actually work with that type, e.g. if the user is not using Float64).

File: src/nonclifford.jl (hunk 1 above)
Comment (Krastanov) [reply src/nonclifford.jl]: I have that expectation because here you are allocating two new intermediary temporary arrays.

Also the `@. iseven(packed_sum)` is a bit weird. This would be more legible as `iseven.(packed_sum)`

File: src/nonclifford.jl (hunk 6 above)
Comment (Fe-r-oz) [reply src/nonclifford.jl:145]: Agreed. 

I asked about whether it is dangerous to modify the dictionary keys while iterating over it. The answer is that it's not!

File: src/nonclifford.jl (hunk 2 above)
Comment (Fe-r-oz) [reply src/nonclifford.jl]: Yeah, the following based on valtype works

```
function _dictvaltype(dict)
    return valtype(dict)
end
```

File: src/nonclifford.jl (hunk 5 above)
Comment (Fe-r-oz) [reply src/nonclifford.jl:158]: Attempt at parameterized version:

```
function apply!(state::GeneralizedStabilizer, gate::PauliChannel; prune_threshold=1e-14)
    dict = state.destabweights
    stab = state.stab
    dtype = _dictvaltype(dict)
    tzero = zero(dtype)
    tone = one(dtype)
    newdict = typeof(dict)(tzero) 

    for ((dᵢ, dⱼ), χ) in dict
        for ((Pₗ, Pᵣ), w) in zip(gate.paulis, gate.weights)
            phaseₗ, dₗ, dₗˢᵗᵃᵇ = rowdecompose(Pₗ, stab)
            phaseᵣ, dᵣ, dᵣˢᵗᵃᵇ = rowdecompose(Pᵣ, stab)
            c = (dot(dₗˢᵗᵃᵇ, dᵢ) + dot(dᵣˢᵗᵃᵇ, dⱼ)) * 2
            dᵢ′ = dₗ ⊻ dᵢ
            dⱼ′ = dᵣ ⊻ dⱼ
            χ′ = χ * w * (-tone)^c * (im)^(-phaseₗ + phaseᵣ + 4)
            newdict[(dᵢ′, dⱼ′)] += χ′
        end
    end

    keys_to_delete = collect(keys(newdict)) 

    for k in keys_to_delete
        if abs(newdict[k]) < prune_threshold
            delete!(newdict, k)
        end
    end

    state.destabweights = newdict
    return state
end
```

File: src/nonclifford.jl (hunk 4 above)
Comment (Fe-r-oz) [reply src/nonclifford.jl:163]: Nope. It's not dangerous.

File: src/nonclifford.jl (hunk 1 above)
Comment (Fe-r-oz) [reply src/nonclifford.jl]: Your approach is much faster!  Also, tried, bitpacking, it's also slower that the current approach. 
```
julia> println("Benchmarking Original Version:")
Benchmarking Original Version:

julia> @btime _allthreesumtozero_original($a, $b, $c)
  10.604 ns (0 allocations: 0 bytes)
false

julia> println("Benchmarking Proposed Version:")
Benchmarking Proposed Version:

julia> @btime _allthreesumtozero_proposed($a, $b, $c)
  2.161 ms (6 allocations: 7.75 MiB)
false
```

File: src/nonclifford.jl (hunk 4 above)
Comment (Fe-r-oz) [reply src/nonclifford.jl:163]: Passed the verdit too soon. 

Another guy provide comprehensive reasons of why it's dangerous and suggested that `filter!` is safe. 'why not use f`ilter!(.., result_dict),` which handles the edge case of rehashing correctly for you' and 'deleting like you're doing is definitely dangerous'

File: src/nonclifford.jl (hunk 6 above)
Comment (Fe-r-oz) [reply src/nonclifford.jl:145]: Another guy provide comprehensive reasons (forwarded them) of why it's dangerous and suggested that `filter!` is safe. 'why not use` filter!(.., result_dict)`, which handles the edge case of rehashing correctly for you' and 'deleting like you're doing is definitely dangerous'

I checked this code passed the test
```
function apply!(state::GeneralizedStabilizer, gate::PauliChannel)
    dict = state.destabweights
    stab = state.stab
    dtype = _dictvaltype(dict)
    tzero = zero(dtype)
    tone = one(dtype)
    newdict = typeof(dict)(tzero) 

    for ((dᵢ,dⱼ), χ) in dict # the state
        for ((Pₗ,Pᵣ), w) in zip(gate.paulis, gate.weights) # the channel
            phaseₗ, dₗ, dₗˢᵗᵃᵇ = rowdecompose(Pₗ, stab)
            phaseᵣ, dᵣ, dᵣˢᵗᵃᵇ = rowdecompose(Pᵣ, stab)
            c = (dot(dₗˢᵗᵃᵇ, dᵢ) + dot(dᵣˢᵗᵃᵇ, dⱼ)) * 2
            dᵢ′ = dₗ .⊻ dᵢ
            dⱼ′ = dᵣ .⊻ dⱼ
            χ′ = χ * w * (-tone)^c * (im)^(-phaseₗ + phaseᵣ + 4)
            
            # Optionally insert into newdict conditionally
            if abs(χ′) >= 1e-14  # TODO parameterize this pruning parameter
                if haskey(newdict, (dᵢ′, dⱼ′))
                    newdict[(dᵢ′, dⱼ′)] += χ′
                else
                    newdict[(dᵢ′, dⱼ′)] = χ′
                end
            end
        end
    end

    # Filter newdict based on abs value
    filter!(x -> abs(x[2]) >= 1e-14, newdict)

    state.destabweights = newdict
    state
end

```

File: src/nonclifford.jl (hunk 5 above)
Comment (Fe-r-oz) [reply src/nonclifford.jl:158]: will add the latest change after filter!

**Comment by Fe-r-oz** [comment]
Kindly please click on the [code review suggestions: TODO2, TODO3](https://github.com/QuantumSavory/QuantumClifford.jl/pull/313/commits/0a1c58f508e07e1b7c06374c6fed85f29f20f5af)[0a1c58f] to have a look at the attempt to TODO2, and TODO3 whenever convenient. 

I sought some advice on this question: "Is this actually necessary? Is it actually dangerous to modify the dictionary keys while iterating over it?" After discussions, it turned out that this is dangerous and` filter! `is safe approach according to folks."

So, I have used the `filter!` and also tried to `parameterize `the prune threshold. The default threshold is `1e-14 `and user can enter their desired thresold. If they don't enter anything, then the default prune threshold is used.  Also, TODO1 is reverted back to what it was as that  is much faster.

Hope this is plausible.

**Comment by Fe-r-oz** [comment]
Kindly please check out a faster approach: 10.6ns to 3.5ns. 

```
julia> using BenchmarkTools

julia> n = 100000; a = rand(Int32, n);b = rand(Int32, n); c = rand(Int32, n);

julia> function _allthreesumtozero(a, b, c)
         @inbounds @simd for i in 1:length(a)
           iseven(a[i] + b[i] + c[i]) || return false
         end
         true
       end
_allthreesumtozero (generic function with 1 method)

julia> function _allthreesumtozero_optimized(a, b, c)
           @inbounds @simd for i in 1:length(a)
               ((a[i] + b[i] + c[i]) & 1) == 0 || return false
           end
           true
       end
_allthreesumtozero_optimized (generic function with 1 method)

julia> println(@benchmark _allthreesumtozero($a, $b, $c))
Trial(10.601 ns)

julia> println(@benchmark _allthreesumtozero_optimized($a, $b, $c))
Trial(3.505 ns)



```

File: src/nonclifford.jl (hunk 2 above)
Comment (Fe-r-oz) [reply src/nonclifford.jl]:
```
julia> d = DefaultDict(0.0im, (falses(1),falses(1))=>1.0+0.0im)
DefaultDict{Tuple{BitVector, BitVector}, ComplexF64, ComplexF64} with 1 entry:
  ([0], [0]) => 1.0+0.0im

julia> valtype(d)
ComplexF64 (alias for Complex{Float64})

julia> eltype(values(d))
ComplexF64 (alias for Complex{Float64})

julia> return eltype(d).parameters[2]
ComplexF64 (alias for Complex{Float64})
```

**Comment by Fe-r-oz** [comment]
I used interactive rebase today by following your email (thanks for comprehensive email) that caused a lot of emails today as I was going through rebase. My apologies for that.

I have also reworded the commits to they are easy to follow. Tested the package locally as well.

**Review by Krastanov — CHANGES_REQUESTED** [review]
Hi, Feroz! Please squash the commits you have made (or clean them up so that one commit is responsible for one task).

For instances, your first commit makes changes to multiple different things that are then undone in other commits, making it difficult to understand what is actually being done:

https://github.com/QuantumSavory/QuantumClifford.jl/pull/313/commits/27953c7b54af4f01e55c3ad02e4905059f517286

Given that the changes here are fairly small, it is probably doable to just squash your commits. If this was a bigger set of changes, it would be best to be careful in separating the commits by topic.

**Comment by Fe-r-oz** [comment]
Closing in favor of #316 

When I tried to squash the commits, the head got diverged and somehow it ended up pushing the commits on my master branch. I tried to reset hard. My apologies for this inconvenience!

Thank you very much again for your comprehensive comments and feedback.

```
pick 867c338 `Destabilizer` now preserves input Stabilizer tableaux
pick 40838f7 `StabMixture` renamed to `GeneralizedStabilizer`
pick 01ea563 rowdecompose was not accounting for input phase
pick 0664b0d `expect` for `GeneralizedStabilizer`
pick e12726d scaffolding for `project!` for `GeneralizedStabilizer`
pick 1f7ad99 Address _dictvaltype, _allthreesumtozero, apply! TODOs
pick f6b6c55 modifying TODO1(_dictvaltype), TODO2(_allthreesumtozero), and TODO3(apply!)
pick 58008c7 fixing tests

# Rebase 0dff482..58008c7 onto 0dff482 (8 commands)
#
# Commands:
# p, pick <commit> = use commit
# r, reword <commit> = use commit, but edit the commit message
# e, edit <commit> = use commit, but stop for amending
# s, squash <commit> = use commit, but meld into previous commit
# f, fixup <commit> = like "squash", but discard this commit's log message
# x, exec <command> = run command (the rest of the line) using shell
# b, break = stop here (continue rebase later with 'git rebase --continue')
# d, drop <commit> = remove commit
# l, label <label> = label current HEAD with a name
# t, reset <label> = reset HEAD to a label
# m, merge [-C <commit> | -c <commit>] <label> [# <oneline>]
# .       create a merge commit using the original merge commit's
# .       message (or the oneline, if no original merge commit was
# .       specified). Use -c <commit> to reword the commit message.
#
# These lines can be re-ordered; they are executed from top to bottom.
```
